File: 변유정/220925_BOJ_1062_가르침.cpp
#include <iostream>
using namespace std;

// 비트마스킹
// 과정
// 1. K(허용되는 알파벳) 범위로 나누기
// 1-1. K 가 5보다 작으면 구현 가능한 단어 없음
// 1-2. K가 26이면 모든 알파벳 사용 가능 (언급한 단어 모두 구현가능)

// 2. 알파벳 사용 배열 만들기
// 2-1. antic 은 기본 사용이므로 미리 표시
// 2-2. 나머지 알파벳은 개수에 맞춰서 선택하기

// 3. 개수에 맞춰서 선택한 알파벳 배열로 구현할 수 있는 단어 최대 개수 찾기
// 3-1. 만약 단어 속 알파벳이 선택되지 않았다면 그 단어는 구현 불가능

// 4. 단어 개수 구했으면 되돌려서 2부터 반복

int N, K;
string words[51];
bool used[26];
int ans = -1;

void dfs(int now, int cnt)
{
	if (cnt + 5 == K)
	{
		int teaching = 0;
		for (int idx = 0; idx < N; idx++)
		{
			bool flag = 0;
			// 범위 좁혀주기 (단어 앞,뒤 4글자는 언제나 동일하다)
			for (int k = 4; k < words[idx].size()-4; k++)
			{
				char tt = words[idx][k];
				if (!used[tt - 'a'])
				{
					flag = 1;
					break;
				}
			}
			if (!flag)
				teaching++;
		}
		if (ans < teaching)
			ans = teaching;
		return;
	}

	for (int i = now; i < 26; i++)
	{
		if (used[i])
			continue;
		used[i] = 1;
		dfs(i + 1, cnt + 1);
		used[i] = 0;
	}
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	cin >> N >> K;
	for (int i = 0; i < N; i++)
	{
		string temp;
		cin >> temp;
		words[i] = temp;
	}

	if (K < 5)
		cout << 0;
	else if (K == 26)
		cout << N;
	else
	{
		string ess = "antic";
		for (int i = 0; i < 5; i++)
			used[ess[i] - 'a'] = 1;
		dfs(0, 0);
		cout << ans;
	}
}

File: 변유정/220925_BOJ_1062_가르침.py
## 여러 방식으로 풀었지만 계속 시간초과나서 잠시 미루기

def choose(now,cnt):
    global ans

    if cnt + 5 == K:
        snt = 0
        for word in words:
            flag = 0
            for apb in word:
                if used[ord(apb) - alpha] == 0:
                    flag = 1
                    break
            if flag == 0:
                snt += 1
        if ans < snt:
            ans = snt
        return

    for i in range(now,26):
        if used[i] == 1:
            continue
        used[i] = 1
        choose(i+1,cnt+1)
        used[i] = 0

N, K = map(int,input().split())
used = [0]*26
alpha = ord('a')
used[ord('a') - alpha] = 1
used[ord('n') - alpha] = 1
used[ord('c') - alpha] = 1
used[ord('i') - alpha] = 1
used[ord('t') - alpha] = 1
words = []
ans = 0
for _ in range(N):
    langs = input()
    langs = langs[4:-4]
    words.append(langs)

if K < 5:
    print(0)
elif K == 26:
    print(N)
else:
    choose(0,0)
    print(ans)